### PR TITLE
{bio}[dummy/dummy] FastQC 0.11.4 + rename fastqc to FastQC + make FastQC 0.10.1 toolchain-neutral (REVIEW)

### PR DESCRIPTION
**Pull request by nathanhaigh**
Working on Easyconfigs for fastqc-0.11.4-foss-2015b

### CODE REVIEW

File: easybuild/easyconfigs/f/fastqc/fastqc-0.11.4-Java-1.8.0_66.eb [hunk 1]
```diff
@@ -0,0 +1,28 @@
+easyblock = 'EB_Java'
```
Comment (boegel) [inline easybuild/easyconfigs/f/fastqc/fastqc-0.11.4-Java-1.8.0_66.eb]: hmm, this is... weird

`EB_Java` is intended to install Java itself, not stuff written in Java...

@nathanhaigh: I know this is not your change, but it just stood out for me, I wonder what this is about

**Comment by boegel** [comment]
Jenkins: ok to test

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
Linux SL 6.7, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6
See https://gist.github.com/9f3036abd13c3b4b45f6 for a full test report.

File: easybuild/easyconfigs/f/fastqc/fastqc-0.11.4-Java-1.8.0_66.eb (hunk 1 above)
Comment (boegel) [inline easybuild/easyconfigs/f/fastqc/fastqc-0.11.4-Java-1.8.0_66.eb]: The only thing that `EB_Java` does on top of `PackedBinary` is to update `$JAVA_HOME`.

Using the `fastqc` install prefix to define `$JAVA_HOME` is probably very wrong?

So, `PackedBinary` should be used here instead (and in other `fastqc` easyconfigs)?

Comment (nathanhaigh) [inline easybuild/easyconfigs/f/fastqc/fastqc-0.11.4-Java-1.8.0_66.eb]: I'll make the changes to this fastqc and then change the older fastqc 
easyconfigs, on which I based this one.

BTW: there are older easyconfigs with different capitalisation (FastQC), 
should I make this consistent across them all?

On 06/01/2016 22:13, Kenneth Hoste wrote:
>
> In easybuild/easyconfigs/f/fastqc/fastqc-0.11.4-Java-1.8.0_66.eb 
> <https://github.com/hpcugent/easybuild-easyconfigs/pull/2267#discussion_r48949343>:
>
> > @@ -0,0 +1,28 @@
> > +easyblock = 'EB_Java'
>
> The only thing that |EB_Java| does on top of |PackedBinary| is to 
> update |$JAVA_HOME|.
>
> Using the |fastqc| install prefix to define |$JAVA_HOME| is probably 
> very wrong?
>
> So, |PackedBinary| should be used here instead (and in other |fastqc| 
> easyconfigs)?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/hpcugent/easybuild-easyconfigs/pull/2267/files#r48949343>.
>



---
This email has been checked for viruses by Avast antivirus software.
https://www.avast.com/antivirus

Comment (boegel) [inline easybuild/easyconfigs/f/fastqc/fastqc-0.11.4-Java-1.8.0_66.eb]: Moving those all in the same directory and consistently applying the fix would be nice, yes (assuming it's the same software). Thanks for looking into this!

**Comment by nathanhaigh** [comment]
Test report by @nathanhaigh
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in this PR)
Linux ubuntu 14.04, Intel(R) Xeon(R) CPU E5-2699 v3 @ 2.30GHz, Python 2.7.6
See https://gist.github.com/121330569fbeb7fba26d for a full test report.

**Comment by nathanhaigh** [comment]
@boegel I'm not sure what to make of this failure.

**Comment by boegel** [comment]
@nathanhaigh: you've removed an existing easyconfig `FastQC-0.10.1-goolf-1.4.10.eb`, which is used as a dependency somewhere else, i.e. in the `Trim_Galore` easyconfig.

You can fix this issue by changing the dependency spec there to:

```python
('FastQC', '0.10.1', '', True)
```

(see also http://easybuild.readthedocs.org/en/latest/Writing_easyconfig_files.html#specifying-dependencies-using-dummy-toolchain)

File: easybuild/easyconfigs/t/Trim_Galore/Trim_Galore-0.3.7-goolf-1.4.10.eb [hunk 2]
```diff
@@ -16,7 +16,7 @@ source_urls = ['http://www.bioinformatics.babraham.ac.uk/projects/%(namelower)s'
 install_cmd = "cp -a %(namelower)s_zip/* %(installdir)s && chmod +x %(installdir)s/%(namelower)s"
 
 dependencies = [
-     ('FastQC', '0.10.1'),
+     ('FastQC', '0.10.1', '', True),
```
Comment (boegel) [inline easybuild/easyconfigs/t/Trim_Galore/Trim_Galore-0.3.7-goolf-1.4.10.eb]: sorry, the `''` should be `-Java-1.7.0_80`, i.e. the versionsuffix you used in the replacement FastQC easyconfig

File: easybuild/easyconfigs/f/FastQC/FastQC-0.10.1-Java-1.7.0_80.eb [hunk 3]
```diff
@@ -17,22 +17,23 @@ description = """FastQC is a quality control application for high throughput seq
...
 
 java = 'Java'
 javaver = '1.7.0_80'
+versionsuffix = '-%s-%s' % (java, javaver)
+dependencies = [(java, javaver)]
 
-dependencies = [('Java', '1.7.0_80', '', True)]
-
-install_cmd = "cp -a FastQC/* %(installdir)s && chmod +x %(installdir)s/fastqc"
+postinstallcmds = ['chmod +x %(installdir)s/fastqc']
```
Comment (boegel) [inline easybuild/easyconfigs/f/FastQC/FastQC-0.10.1-Java-1.7.0_80.eb]: why was the `cp` removed here @nathanhaigh? this changes the way it was installed before (cfr. the changes in the sanity check)

are the other files useless? I see `.jar`s in there that may be needd?

Comment (boegel) [inline easybuild/easyconfigs/f/FastQC/FastQC-0.10.1-Java-1.7.0_80.eb]: ping?

Comment (nathanhaigh) [inline easybuild/easyconfigs/f/FastQC/FastQC-0.10.1-Java-1.7.0_80.eb]: @boegel - Rather than defining a custom `install_cmd` I'm using the default install action of a `PackedBinary` which is to copy all the contents of the unpacked binary to the `installdir`. All, that is required is to then do a simple post-install fix of executable permissions on the `fastqc` file.

So the result should be the same - are you seeing something different?

**Comment by nathanhaigh** [comment]
I see that the main `fastqc` executable is in fact a Perl script. Perl should then be added as a dependency and the shebang line patched.

**Comment by nathanhaigh** [comment]
@boegel The merge conflict I assume is due to recent changes in `develop` which correctly modified the toolchain to `dummy`. These files have since been moved into a single `FastQC` directory and had the Java version suffix appended to the filenames.

I'm not sure how to go about resolving this conflict myself.

**Comment by boegel** [comment]
@nathanhaigh: conflicts with `develop` fixed in https://github.com/nathanhaigh/easybuild-easyconfigs/pull/2

they were fairly trivial, mostly because of style fixes done recently (cfr. #2492)

**Comment by boegel** [comment]
@nathanhaigh: having a test to check for hardcoded paths in shebang lines makes sense... Please open a separate issue on that?

Are you going to look into adding a `Perl` dep for each of these?
In that case, I'm not sure how wise it is to stick to the `dummy` toolchain...

File: easybuild/easyconfigs/f/FastQC/FastQC-0.10.1-Java-1.7.0_80.eb [hunk 4]
```diff
@@ -17,22 +17,25 @@ description = """FastQC is a quality control application for high throughput seq
...
 
 sanity_check_paths = {
-    'files': ["fastqc", "fastqc_icon.ico", "INSTALL.txt", "jbzip2-0.9.jar", "LICENSE.txt",
-                "README.txt", "RELEASE_NOTES.txt", "run_fastqc.bat", "sam-1.32.jar"],
-    'dirs': ["Contaminants", "Help", "Templates", "uk"],
+    'files': ['fastqc'],
```
Comment (boegel) [inline easybuild/easyconfigs/f/FastQC/FastQC-0.10.1-Java-1.7.0_80.eb]: why did you change this?

File: easybuild/easyconfigs/f/FastQC/FastQC-0.10.1-Java-1.7.0_80.eb (hunk 3 above)
Comment (boegel) [inline easybuild/easyconfigs/f/FastQC/FastQC-0.10.1-Java-1.7.0_80.eb]: Well, no, but I then wonder why you changed the `sanity_check_paths` too; I assumed you had to do that because of this change, but if `PackedBinary` copies everything, there's no reason to change `sanity_check_paths`?

File: easybuild/easyconfigs/f/FastQC/FastQC-0.10.1-Java-1.7.0_80.eb [hunk 5]
```diff
@@ -17,22 +17,31 @@ description = """FastQC is a quality control application for high throughput seq
...
 
 sanity_check_paths = {
+<<<<<<< HEAD:easybuild/easyconfigs/f/FastQC/FastQC-0.10.1-Java-1.7.0_80.eb
+    'files': ['fastqc'],
+    'dirs': [],
+=======
     'files': ["fastqc", "fastqc_icon.ico", "INSTALL.txt", "jbzip2-0.9.jar", "LICENSE.txt",
               "README.txt", "RELEASE_NOTES.txt", "run_fastqc.bat", "sam-1.32.jar"],
     'dirs': ["Contaminants", "Help", "Templates", "uk"],
+>>>>>>> develop:easybuild/easyconfigs/f/FastQC/FastQC-0.10.1-goolf-1.4.10.eb
```
Comment (boegel) [inline easybuild/easyconfigs/f/FastQC/FastQC-0.10.1-Java-1.7.0_80.eb]: merge gone wrong...

can you look into fixing this @nathanhaigh?

may be my bad, let me know if you need help with this

File: easybuild/easyconfigs/f/FastQC/FastQC-0.10.1-Java-1.7.0_80.eb (hunk 3 above)
Comment (nathanhaigh) [inline easybuild/easyconfigs/f/FastQC/FastQC-0.10.1-Java-1.7.0_80.eb]: @boegel I think I was aiming to simplyfy by just having the check on the executable. Shall I reinstate all the files/dirs for all the easyconfigs?

Comment (boegel) [inline easybuild/easyconfigs/f/FastQC/FastQC-0.10.1-Java-1.7.0_80.eb]: well, maybe checking for some of them isn't very useful, but I'd say the `.jar` files are?

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 3 out of 6 (6 easyconfigs in this PR)
node2090.delcatty.os - Linux SL 6.7, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6
See https://gist.github.com/67ae4055549f4b6ae446 for a full test report.

File: easybuild/easyconfigs/f/FastQC/FastQC-0.10.1-Java-1.7.0_80.eb [hunk 6]
```diff
@@ -17,22 +17,26 @@ description = """FastQC is a quality control application for high throughput seq
...
 
-dependencies = [('Java', '1.7.0_80', '', True)]
-
-install_cmd = "cp -a FastQC/* %(installdir)s && chmod +x %(installdir)s/fastqc"
+postinstallcmds = ['chmod +x %(installdir)s/fastqc']
 
 sanity_check_paths = {
     'files': ["fastqc", "fastqc_icon.ico", "INSTALL.txt", "jbzip2-0.9.jar", "LICENSE.txt",
-              "README.txt", "RELEASE_NOTES.txt", "run_fastqc.bat", "sam-1.32.jar"],
+              "README.txt", "RELEASE_NOTES.txt", "run_fastqc.bat", "sam-1.32.jar", "cisd-jhdf5.jar"],
```
Comment (boegel) [inline easybuild/easyconfigs/f/FastQC/FastQC-0.10.1-Java-1.7.0_80.eb]: @nathanhaigh copy-paste error, the `.jar` you added isn't there yet for this FastQC version?

File: easybuild/easyconfigs/f/FastQC/FastQC-0.11.4-Java-1.8.0_74.eb [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+javaver = '1.8.0_74'
+versionsuffix = '-%s-%s' % (java, javaver)
+dependencies = [(java, javaver)]
+
+postinstallcmds = ['chmod +x %(installdir)s/fastqc']
+
+sanity_check_paths = {
+    'files': ["fastqc", "fastqc_icon.ico", "INSTALL.txt", "jbzip2-0.9.jar", "LICENSE.txt",
+              "README.txt", "RELEASE_NOTES.txt", "run_fastqc.bat", "sam-1.32.jar", "cisd-jhdf5.jar"],
+    'dirs': ["Contaminants", "Help", "Templates", "uk"],
```
Comment (boegel) [inline easybuild/easyconfigs/f/FastQC/FastQC-0.11.4-Java-1.8.0_74.eb]: why isn't this the same as in the other FastQC 0.11.4 easyconfig?

Comment (boegel) [inline easybuild/easyconfigs/f/FastQC/FastQC-0.11.4-Java-1.8.0_74.eb]: it's failing btw like this (see test report)

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 2 out of 6 (6 easyconfigs in this PR)
node2090.delcatty.os - Linux SL 6.7, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6
See https://gist.github.com/6968be46a69678b0a013 for a full test report.

File: easybuild/easyconfigs/f/fastqc/fastqc-0.11.2.eb [hunk 8]
```diff
@@ -1,27 +0,0 @@
...
-description = """A set of tools (in Java) for working with next generation sequencing data in the BAM format."""
-
-toolchain = {'name': 'dummy', 'version': ''}
-
-source_urls = ['http://www.bioinformatics.babraham.ac.uk/projects/fastqc/']
-sources = ['%(name)s_v%(version)s.zip']
-
-dependencies = [('Java', '1.7.0_60')]
-
-postinstallcmds = ["chmod -R a+rx %(installdir)s/fastqc %(installdir)s/run_fastqc.bat"]
```
Comment (boegel) [inline easybuild/easyconfigs/f/fastqc/fastqc-0.11.2.eb]: the `chmod` on `run_fastqc.bat` got removed?

File: easybuild/easyconfigs/f/fastqc/fastqc-0.11.2.eb [hunk 9]
```diff
@@ -1,27 +0,0 @@
...
-
-modextrapaths = {'PATH': ['']}
```
Comment (boegel) [inline easybuild/easyconfigs/f/fastqc/fastqc-0.11.2.eb]: @nathanhaigh: without this, the `fastqc -v` sanity check command will fail

**Comment by boegel** [comment]
This has taken long enough, so I got this merged via #2657 to avoid losing more time on it...

Thanks for your efforts @nathanhaigh!

File: easybuild/easyconfigs/f/fastqc/fastqc-0.11.2.eb (hunk 8 above)
Comment (nathanhaigh) [inline easybuild/easyconfigs/f/fastqc/fastqc-0.11.2.eb]: @boegel - There is no need for this to be executable. It's intended to be a Windows batch file and the `fastqc` executable is inteneded for Linux.

Also, the `fastqc` executable is actually a Perl script and Perl isn't included as a dependency. However, it is just a wrapper for processing command line arguments and calling java, so maybe a hard dependency isn't required.

Comment (boegel) [inline easybuild/easyconfigs/f/fastqc/fastqc-0.11.2.eb]: I checked the contents, and it doesn't seem to be Windows-specific to me?

The name does suggest so, that's true.

File: easybuild/easyconfigs/f/fastqc/fastqc-0.11.2.eb (hunk 9 above)
Comment (nathanhaigh) [inline easybuild/easyconfigs/f/fastqc/fastqc-0.11.2.eb]: @boegel - I thought the easyblock `Binary` which `PackedBinary` inherits, sets this by default?

Comment (boegel) [inline easybuild/easyconfigs/f/fastqc/fastqc-0.11.2.eb]: ah, could be, damn, I should have checked that... :(

Shall I issue a follow-up PR to remove this again?

File: easybuild/easyconfigs/f/fastqc/fastqc-0.11.2.eb (hunk 8 above)
Comment (nathanhaigh) [inline easybuild/easyconfigs/f/fastqc/fastqc-0.11.2.eb]: According to their docs http://www.bioinformatics.babraham.ac.uk/projects/fastqc/INSTALL.txt:

> Running FastQC Interactively
----------------------------
Windows: Simply double click on the `run_fastqc.bat` file.  If you want to make a pretty 
shortcut then we've included an icon file in the top level directory so you don't have
to use the generic bat file icon.

File: easybuild/easyconfigs/f/fastqc/fastqc-0.11.2.eb (hunk 9 above)
Comment (nathanhaigh) [inline easybuild/easyconfigs/f/fastqc/fastqc-0.11.2.eb]: Since the easyblocks are supposed to simplify these things, I would remove it.
